# Linear Algebra

In [1]:
import numpy as np

The numpy.trace() function is used to return the sum of the diagonals of the matrix.

In [4]:
matrix = np.array([[1,2],
                   [4,6]])

result = np.trace(matrix)

print(result)

7


In [5]:
matrix = np.array([[10,2,3],
                   [4,5,2],
                   [2,2,1]])

matrix_trace = np.trace(matrix)

print("trace of matrix = ", matrix_trace)

trace of matrix =  16


## The determinant

In [7]:
A = np.array([[1,2],[3,4]])
res = np.linalg.det(A)
print(res)

-2.0000000000000004


In [8]:
#Another example
import numpy as np

# Define a square matrix
matrix = np.array([[1, 2], [4, 3]])

# Calculate determinant
determinant = np.linalg.det(matrix)

print(determinant)

-4.999999999999999


In [9]:
if A.shape[0] == A.shape[1]:
    print('matrix is a square')
else:
    print('Matrix is not a square')

matrix is a square


# Inverse matrix
When Is a Matrix Invertible?
Before jumping into code, it’s crucial to know this: not all matrices have an inverse. A matrix is invertible (or non-singular) if:

1. It is a square matrix (same number of rows and columns)
2. Its determinant is not zero

In [10]:
# Step 1: Create a square matrix
A = np.array([[4,7],
              [2,6]])

# Step 2: Calculate its inverse
A_inv = np.linalg.inv(A)

print('Original Matrix A:')
print(A)

print('Inverse of A:')
print(A_inv)

Original Matrix A:
[[4 7]
 [2 6]]
Inverse of A:
[[ 0.6 -0.7]
 [-0.2  0.4]]


In [ ]:
import pandas as pd

Let’s verify our inverse calculation by multiplying the original and its inverse. The result should be close to an identity matrix (within floating point precision):

In [14]:
identity_check = np.dot(A, A_inv)
print('The product of A and its inverse should be the identity matrix:')
print(identity_check)

The product of A and its inverse should be the identity matrix:
[[ 1.00000000e+00 -1.11022302e-16]
 [-1.11022302e-16  1.00000000e+00]]


If your matrix is not invertible (e.g., its determinant is zero), NumPy will throw a LinAlgError

In [15]:
# Non-invertible matrix
B = np.array([[1, 2],
              [2, 4]])

B_inv = np.linalg.inv(B)

LinAlgError: Singular matrix

Always check the determinant first before attempting an inversion:

In [16]:
if np.linalg.det(B) != 0:
    B_inv = np.linalg.inv(B)
else:
    print('Matrix is singular and cannot be inverted')

Matrix is singular and cannot be inverted


### Solving Linear Equations

In [17]:
#Case A: m=n
# A 2 x 2 matrix
A = np.array([[2,3],
              [4,6]])
b = np.array([5,10])
# A.shape[1] gives the number of columns of A
# Solution

det_A = np.linalg.det(A)

def solution(A, b):
    if np.linalg.det(A) != 0:
        inv = np.inalg.inv(A)
        sol = inv@B
        return sol
    else:
        return 'infinite solution or no solution'
print('Determinant of A is ', det_A)
solution(A,b)

Determinant of A is  0.0


'infinite solution or no solution'

Solve Using numpy.linalg.solve()

In [18]:
x = np.linalg.solve(A, b)
print('Solution', x)

LinAlgError: Singular matrix

Solving a system with a unique solution

In [19]:
A = np.array([[2, 3],
              [3, 4]])
B = np.array([8, 11])
x = np.linalg.solve(A, B)
print("Solution:", x)

Solution: [1. 2.]


In [20]:
#How do we know that x = 1 and y = 2 are correct? We can multiply the matrix A with the result x and compare it to B:

print('Check', np.dot(A, x))

Check [ 8. 11.]


# Understanding Eigenvalues and Eigenvectors
In linear algebra, eigenvalues and eigenvectors play a crucial role in decomposing matrices, solving systems of equations, and even powering algorithms in machine learning and computer graphics.

In [26]:
A = np.array([[4,2],[1,3]])

We're working with a 2x2 matrix here. Eigen decomposition is only defined for square matrices, so always verify the matrix shape using:

In [27]:
print('Matrix shape:', A.shape)

Matrix shape: (2, 2)


Compute Eigenvalues and Eigenvectors

In [28]:
eigenvalues, eigenvectors = np.linalg.eig(A)

Explanation: np.linalg.eig() returns two results:

eigenvalues: A 1D array of eigenvalues.
eigenvectors: A 2D array where each column is an eigenvector corresponding to an eigenvalue.

In [29]:
print("Eigenvalues:", eigenvalues)
print("Eigenvectors:", eigenvectors)

Eigenvalues: [5. 2.]
Eigenvectors: [[ 0.89442719 -0.70710678]
 [ 0.4472136   0.70710678]]


In [30]:
np.linalg.eig(A)

EigResult(eigenvalues=array([5., 2.]), eigenvectors=array([[ 0.89442719, -0.70710678],
       [ 0.4472136 ,  0.70710678]]))

Therefore, the matrix A has two eigenvalues: 5 and 2. Each one has a corresponding eigenvector:

The first eigenvector [0.894, 0.447] is scaled by 5 when multiplied by the matrix A.

The second eigenvector [-0.707, 0.707] is scaled by 2.

Verify the Eigenvalue Equation:

The property of eigenvectors is:

A * v = λ * v

Let’s verify this manually for one eigenvalue-eigenvector pair:


In [31]:
# Pick the first eigenvector and eigenvalue
v = eigenvectors[:,0]
λ = eigenvalues[0]

Av = A @ v
λv = λ *v


print("A @ v:", Av)
print("λ * v:", λv)
#Expected Output: The two results should be approximately equal 
#(allowing for minor floating-point differences).


A @ v: [4.47213595 2.23606798]
λ * v: [4.47213595 2.23606798]


#  Diagonalization
A matrix A is diagonalizable if:

A = P * D * P^-1
Where:

P is the matrix of eigenvectors
D is the diagonal matrix of eigenvalues.

Let's Perform Diagonalization

In [32]:
P = eigenvectors
D = np.diag(eigenvalues)
P_inv = np.linalg.inv(P)

A_reconstructed = P @ D @ P_inv

print("Reconstructed A:", A_reconstructed)
print("Original A:", A)
# If both are equal the the matrix A is diagonalizable

Reconstructed A: [[4. 2.]
 [1. 3.]]
Original A: [[4 2]
 [1 3]]


Use np.allclose() instead of == for floating-point equality checks

In [33]:
print("Is reconstructed A close to original A?", np.allclose(A, A_reconstructed))

Is reconstructed A close to original A? True


Recall that only square matrices can be diagonalized.

By understanding diagonalization, you gain the tools to simplify matrix operations and peek deeper into the nature of systems and data structures.